internal/trace/transport.go

```diff
@@ -53,35 +54,42 @@ func (t *Transport) RoundTrip(req *http.Request) (resp *http.Response, err error
 	ctx := req.Context()
 	e := log.GetLogger(ctx)
 
-	e.Debugf("> Request: %q %q", req.Method, req.URL)
-	e.Debugf("> Request headers:")
-	logHeader(req.Header, e)
+	// logs to be printed out
+	var logs strings.Builder
+	fmt.Fprintf(&logs, "> Request: %q %q\n", req.Method, req.URL)
+	fmt.Fprintln(&logs, "> Request headers:")
+	fmt.Fprint(&logs, logHeader(req.Header))
 
 	resp, err = t.RoundTripper.RoundTrip(req)
 	if err != nil {
+		e.Debug(logs.String())
 		e.Errorf("Error in getting response: %w", err)
 	} else if resp == nil {
+		e.Debug(logs.String())
 		e.Errorf("No response obtained for request %s %q", req.Method, req.URL)
 	} else {
-		e.Debugf("< Response status: %q", resp.Status)
-		e.Debugf("< Response headers:")
-		logHeader(resp.Header, e)
+		fmt.Fprintf(&logs, "< Response status: %q\n", resp.Status)
+		fmt.Fprintln(&logs, "< Response headers:")
+		fmt.Fprint(&logs, logHeader(resp.Header))
+		e.Debug(logs.String())
 	}
 	return resp, err
 }
```

Comment:
This logging mechanism does not preserve calling order.

Suppose we have the following workflow,
- T0: Send request 1
- T1: Send request 2
- T2: Receive response for request 2
- T3: Receive response for request 1

In this case, request 1 is sent before request 2. However, in the log, request 2 is logged before request 1, which is not correct.

Reply:
You may take a look at https://github.com/oras-project/oras/blob/main/internal/trace/transport.go and see how oras resolves it.

Reply:
Updated.